Creates Dictionaries of Facets for an Object in Advanced Search

angcast opened the pull request:
Method creates a dictionary called dict_facets. The keys of dict_facets are the facets of the objects in advanced search. The keys’ value is a list containing tuples of the top 10 most occuring instances of the facet. The format of the tuples is (instance, number of occurrences). 

Example of Organization:

dict_facets={
            "authors":  [("Philip M. Morse”,1)]
            “author organizations”: [(“MIT”,1),("USA",2)]
	     “recipients”: [(“N. M. Sage”,1)]
	     “recipient organizations”: [("Committee on Machine Computation”,1)]
	      “cceds”: [(“J. W. Forrester”,1)]
	      “cced organizations”: [(“LWH”,1)]
               "years": [(1950,1)]
  }

How to access:

dict_facets["authors"] returns [("Philip M. Morse",1)]
dict_facets["authors"][0] returns "Philip M. Morse"
dict_facets["authors"][1] returns 1

When a user searches for something, we can display these facets. These facets can be cool to look at; since, it gives a big overview of the documents. It will show what instances of the facets were most relevant depending on the user's search. Also, it will be easier to see patterns in our documents. In addition, we can use the code as a foundation for when we create an option to sort the documents in order of relevance.

Carol217 commented on djweb/dj_comp_hist/views.py:
```diff
@@ -179,13 +179,22 @@ def browse(request):
+    counter_cceds=Counter()
+    counter_cced_organizations=Counter()
     for doc in doc_objs.all():
         counter_dates[doc.date.year]=+1
```
This error was thrown: 

~\Documents\GitHub\computation_hist\djweb\dj_comp_hist\views.py in generate_search_facets(doc_objs)
    213     counter_cced_organizations=Counter()
    214     for doc in doc_objs.all():
--> 215         counter_dates[doc.date.year]=+1
    216         for author in doc.author_person.all():
    217             counter_authors[author.fullname]=+1

AttributeError: 'NoneType' object has no attribute 'year'

There's no object called year; you'll have to find a different way to parse it.

angcast replied:
Hey Carol, that's super weird because I tested the code using embed and the advanced search. I was wondering what you typed in the shell? So, I could fix it.

srisi replied:
Some dates are None. I guess that makes sense (they are undated). 
You can just catch the error:
```
# Some dates are None -> skip those documents
if doc.date:
    counter_dates[doc.date.year] += 1
```

Carol217 commented on djweb/dj_comp_hist/views.py:
```diff
@@ -197,17 +206,18 @@ def generate_search_facets(doc_objs):
+                 "cced organizations":dict(counter_cced_organizations.most_common(10)),
+                 "years": dict(counter_dates.most_common(10)),
+                 }
+    return dict_facets
```
a little more spacing would be nice here. if you could look for the yellow squiggly lines and follow their conventions that'd be nice! otherwise no pressure

Carol217 commented on djweb/dj_comp_hist/views.py:
```diff
@@ -179,13 +179,22 @@ def browse(request):
+    Creates a dictionary with dictionaries of each facet of the given parameter. Each of the facet dictionaries contain
+    keys that are the top 10 most common instances of that given facet and values that signify the number of
+    occurrences of that specific instance.
+
```
the way I ran this (might be useful to include in documentation)
In terminal:
```
cd djweb
python manage.py shell
```
In shell:
```
from dj_comp_hist.models import Document
doc_objs = Document.objects
from dj_comp_hist.views import generate_search_facets
generate_search_facets(doc_objs)
```

mscuthbert commented:
Cool -- can you say what aspects of advanced search these `dict_facets` will help with.  Thanks!

ryaanahmed commented:
@angcast - just checking in to see what the status of this PR is. Is this still something you're working on? If so, would love to see the style changes @Carol217 requested implemented, and also some further explanation per @mscuthbert's ask above. Thanks!

angcast commented:
Yes, I am still working on it. I got sick right before spring break and was not able to complete it. I apologize. However, since I am not sick anymore, I am going to finish it up and implement it!  @ryaanahmed . I will probably have it done during my UROP hours.

angcast commented:
Hi, I fixed the issue Carol was talking about. Also, I created a display for the facets. In addition, I updated the description @Carol217 @mscuthbert  . I created a foundation for the display of these facets. I think it would be a cool thing to display. I was wondering if you guys think the same. If so, I can create a new issue regarding this and work on it.

ryaanahmed commented on templates/archives/advanced_search.jinja2:
```diff
@@ -130,6 +130,57 @@
             {% endif %}
         </div>
+        <div class="col-lg-3 col-md-3 col-sm-12">
+            <label class="text_label" for="title">Authors:</label>
```
You've also got some </label> tags but without the corresponding opening <label> tag. e.g.,

```
Author: </label> {{facet[0]}}
```

mscuthbert commented:
@ryaanahmed -- on Friday we were wondering how you can check the number of database queries being done.  If this is something to write up later this week, that would help people understand prefetch_related's role.  Thanks!

angcast commented:
Thank you for this! I wanted to test my code, but I am still having problems with the splite3. Do you know when we are removing it?

ryaanahmed commented:
@angcast - we're not removing sqlite3, but once we merge in #280 , we will no longer need the fts4 module that is not working on your system.

Did you manage to get that dll replaced everywhere? Any luck with it, or still getting the same error?

ryaanahmed commented:
@angcast - #280 is now in, which removes the module that wasn't working on your machine. Try merging from master and seeing if you can get populate_from_metadata working so you can test your code locally.

I played with this today, and with the prefetch the actual counter construction doesn't take much time—especially on a smaller subset of the documents (i.e., when an actual search is run, not just hitting search with no inputs). I'd say once you get the template a little better cleaned up, this will be ready to merge, and we can talk about next steps, like making the visualization of the facets more interesting.

angcast commented:
Yay it works! Thank you

ryaanahmed commented:
@angcast — I've fixed the issues in the template and merged this work into our `unify_search` branch, as it's now different enough from the version of search that this was based on that things were about to lead to a very messy merge! Closing this, and we'll include this feature in our PR for that work.
